Add material icons for file list

wxiaoguang opened the pull request:
Close #11149

Co-authored-by: Yarden Shoham <git@yardenshoham.com>


![image](https://github.com/go-gitea/gitea/assets/2114189/09fc4371-5f85-445a-bd05-a89bc3fbae96)

yardenshoham reviewed (commented):
I had to add 4 icons last time to complete the set. I had them in `web_src/svg/`. You might want them too.

In my implementation, it looked like

#### Before
![before](https://user-images.githubusercontent.com/20454870/232415994-5420baee-df5c-4ee2-a5d0-a504730252a9.png)
#### After
![after](https://user-images.githubusercontent.com/20454870/232416184-6b4528a5-4f01-4996-baf9-da5b2d2d0ff5.png)

### material-file.svg
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="svg material-file" width="16" height="16" aria-hidden="true"><path fill="#90a4ae" d="M13 9h5.5L13 3.5V9M6 2h8l6 6v12a2 2 0 0 1-2 2H6a2 2 0 0 1-2-2V4c0-1.11.89-2 2-2m5 2H6v16h12v-9h-7V4z"/></svg>
```

The files:

### material-folder-root.svg
```svg
<svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" class="svg material-folder-root" width="16" height="16" aria-hidden="true"><path fill="#90a4ae" d="M12 20a8 8 0 0 1-8-8 8 8 0 0 1 8-8 8 8 0 0 1 8 8 8 8 0 0 1-8 8m0-18A10 10 0 0 0 2 12a10 10 0 0 0 10 10 10 10 0 0 0 10-10A10 10 0 0 0 12 2m0 5a5 5 0 0 0-5 5 5 5 0 0 0 5 5 5 5 0 0 0 5-5 5 5 0 0 0-5-5z"/></svg>
```

### material-folder-symlink.svg
```svg
<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M10 4H4c-1.11 0-2 .89-2 2v12a2 2 0 002 2h16a2 2 0 002-2V8a2 2 0 00-2-2h-8l-2-2z" fill="#42a5f5" opacity=".745"/><path d="M16.972 10.757v2.641h-6.561v5.281h6.561v2.641l6.562-5.281-6.562-5.282z" opacity=".81" fill="#c5e5fd"/></svg>
```

### material-folder.svg
```svg
<svg xmlns="http://www.w3.org/2000/svg" width="24" height="24" viewBox="0 0 24 24"><path d="M10 4H4c-1.11 0-2 .89-2 2v12a2 2 0 002 2h16a2 2 0 002-2V8a2 2 0 00-2-2h-8l-2-2z" fill="#42a5f5"/></svg>
```

delvh commented:
What I haven't quite understood yet: What's the difference to #24147?
Wasn't its main problem that it should be configurable?

yardenshoham commented:
> Wasn't its main problem that it should be configurable?

I think here you can untar, change files, tar again.

yardenshoham commented:
Something bad happened here

![image](https://github.com/go-gitea/gitea/assets/20454870/0caf8db8-b3a2-42c0-a10c-fe6fc52fbc30)

I remember that what fixed that for me was passing the SVGs through `build/generate-svg.js`'s `processFile`

silverwind commented:
I think I prefer octicons for unrecognized folders and files and a option could be added that enables the file type icons. I personally would prefer for folders to not receive any custom icons.

lunny commented on modules/fileicon/basic.go:
```diff
@@ -0,0 +1,33 @@
+}
+
+func FileIcon(ctx context.Context, entry *git.TreeEntry) template.HTML {
+	// TODO: if it needs to use different file icon provider for different users, it could use ctx to check user setting and call fileIconBasic(ctx, entry)
```
I think at least, this PR keeps the freedom of changing the file icons of the whole instance. Per user configurations could be next PR.

wxiaoguang replied:
I think we can introduce an "undocumented" config option in this PR to set "basic" or "material" icon set, then replace the option with user setting in next PR soon.

Since 1.21 is still at early stage, the undocumented option won't cause end-user problems IMO.

lunny commented:
Could we have a configuration to allow users to switch back to `Basic` from the default `Material`? Then I will give my L-G-T-M

puni9869 commented:
Will it be the default look of icon in 1.21
Its looks good.

wxiaoguang commented:
> What I haven't quite understood yet: What's the difference to #24147? Wasn't its main problem that it should be configurable?

The differences:

* This PR uses LFS to store the icon pack file (.tgz), so only 300+ lines are added (the original one added more than 6000 lines)
    * The LFS could also benefit other large file storing in Gitea's git repo in the future, to avoid bloating the main code base.
* The external icon pack doesn't go into Gitea's image assets, then they can be modified/removed independently.
* End users can make/use/publish their own icon pack
    * It's possible to support more "plug-able pack files"

JakobDev commented:
Sound code, but does this work with different styles? It is possible to select a style in Gitea. Does the Icons look good on every style?

wxiaoguang commented:
> Sound code, but does this work with different styles? It is possible to select a style in Gitea. Does the Icons look good on every style?

The quick is answer is: it might not work that well. Because there are some "light" icons in the icon set, but they are not used yet. But I think it's not in this PR's scope. Due to how Gitea's backend/frontend work, it needs a lot of more time to fix it.

wxiaoguang commented:
@yardenshoham 

> I had to add 4 icons last time to complete the set. I had them in web_src/svg/. You might want them too.
> In my implementation, it looked like .....

If I understand correctly, only "material-folder(-generic).svg" and "material-folder-symlink.svg" are used, now they are there.

> Something bad happened here
> I remember that what fixed that for me was passing the SVGs through build/generate-svg.js's processFile

There were also a lot of backend code handling SVG contents, which were fragile. Now I have rewritten them (and add tests).


And I also added a temporary option `[ui].FILE_ICON_THEME` , in case some edge versions users would like to revert to the old icon theme.

yardenshoham commented:
The icons look a bit small and too high
![image](https://github.com/go-gitea/gitea/assets/20454870/d9343b6f-c2d5-4cdd-9cb3-74fde04670cc)

wxiaoguang commented:
It could be an old bug

try.gitea.io:

![image](https://github.com/go-gitea/gitea/assets/2114189/185a36df-8c34-4cdb-a235-a81989a06496)

new:

![image](https://github.com/go-gitea/gitea/assets/2114189/02c5ede1-d2a5-4734-ac0a-429e56e7edbd)

yardenshoham commented:
Look in my screenshot in .gitignore and .gitpod.yml, are you sure you want it this way?

wxiaoguang commented:
> Look in my screenshot in .gitignore and .gitpod.yml, are you sure you want it this way?

I don't want it this way, but it's a longstanding problem (you can also check https://user-images.githubusercontent.com/20454870/232415221-84411fb8-c154-4a11-b523-226fa7ce0842.png in #24147 ) . That's not a new problem.

![image](https://github.com/go-gitea/gitea/assets/2114189/3e3f2eac-f08c-4e15-b5dc-0fad7435fab5)

----

So, to avoid mixing too many things together, the SVG alignment problem could be fixed in a separate PR.

yardenshoham reviewed (approved):
I like it, good idea

silverwind commented:
I think this needs a bit more discussion:

- The script seems to pull from npm, why not add the icons as a regular npm dependency in package.json instead?
- Can we make the blue icons follow `--color-primary`?

yardenshoham commented:
Regarding npm, I think we want to keep it as an archive to reduce space on disk

yardenshoham commented:
But then it's not customizable

wxiaoguang commented:
> * The script seems to pull from npm, why not add the icons as a regular npm dependency in package.json instead, which brings the benefit of locking the dependency version?

It depends, I do not consider it as a real "dependency", just like the "gitignore" and "license" files, these resrouces are just fetched from external sites.

While I do not have objection for using npm, as long as the management and usage is clear.

> * Can we make the blue icons follow `--color-primary`?

It depends, but it's not necessary from my side.

> * Isn't the introduction of LFS for a single tgz file excessive? Can gitea repo still be mirrored to servers that have LFS disabled?

IMO it's better than managing hundreds external asset files.

IIRC at the time of #24147 there were about 600 SVG files, but now it has more than 800 files. Keeping updating these files into Gitea's code base, then it becomes another "Golang vendor" directory problem.

silverwind commented:
I'd be fine to check in the tgz file into the repo as is, but requiring LFS on both client on server for this repo seems suboptimal. We don't even document this new LFS dependency and in the past I have been avoiding LFS precicely because it is not part of a standard git installation.

wxiaoguang commented:
Committing a binary file into git repo as regular file is also sub-optimal. When the file changes, the "diff" algorithm doesn't work, every file is stored as a new one with its 100% content.

If most people like storing it as a regular file, or like storing separate files without archive, or using the current LFS approach,  either is fine to me. It depends the majority preference.

wxiaoguang commented:
There could be another choice: do not store the files into Gitea's repo, but provide them as a downloadable asset pack.

silverwind commented:
My preferred solution would be to check the tgz it into regular git, generated via script from a npm dependency.

KN4CK3R commented:
My preferred solution would be no tgz in git at all. Just the ability to add these icon packs in a folder so that a user can choose one "theme". The default would be no icon pack and the generic icons like we have at the moment. We could offer icon packs at `https://dl.gitea.com/gitea_icons/` or similar.

yardenshoham commented:
I prefer material as the default

silverwind commented:
> My preferred solution would be no tgz in git at all. Just the ability to add these icon packs in a folder so that a user can choose one "theme". The default would be no icon pack and the generic icons like we have at the moment. We could offer icon packs at `https://dl.gitea.com/gitea_icons/` or similar.

I think we should provide ship least one default pack. I have not looked at this pack, but are there existing other packs that work? Maybe existing VSCode icon packs?

lunny commented:
> > My preferred solution would be no tgz in git at all. Just the ability to add these icon packs in a folder so that a user can choose one "theme". The default would be no icon pack and the generic icons like we have at the moment. We could offer icon packs at `https://dl.gitea.com/gitea_icons/` or similar.
> 
> I think we should provide ship least one default pack. I have not looked at this pack, but are there existing other packs that already work? Maybe we could have existing VSCode icon packs compatible?

The default basic theme kept as before.

wxiaoguang commented on custom/conf/app.example.ini:
```diff
@@ -1207,6 +1207,9 @@ LEVEL = Info
 ;THEMES = auto,gitea,arc-green
 ;;
+;; The icons for file list (basic/material), this is a temporary option which will be replaced by a user setting in the future.
+;FILE_ICON_THEME = material
```
I think I have explained the decision above.

https://github.com/go-gitea/gitea/pull/26236#issuecomment-1657791296

![image](https://github.com/go-gitea/gitea/assets/2114189/e85338d6-7f53-471f-bcb0-78e450877a4f)

wxiaoguang commented on modules/html/html.go:
```diff
@@ -6,28 +6,20 @@ package html
-		if defaultClass == "" {
-			class = _class
-		} else {
-			class = defaultClass + " " + _class
+	if len(others) >= 2 {
+		if v, ok := others[1].(string); ok && v != "" {
+			if class != "" {
+				class += " "
+			}
+			class += v
```
That's the old behavior

yardenshoham commented:
@silverwind do you want to block?

yardenshoham commented:
How do you customize with option 3?

silverwind commented:
I still don't like the introduction of LFS for this, it means the gitea repo can no longer be hosted on a default installation of gitea which has LFS disabled. 294kB is certainly not a size where I would even consider LFS, that starts in the realms of 100MB+ for me.

delvh commented:
> How do you customize with option 3?

Isn't that even the easiest one to customize?
You can simply add/change/remove files in the directory to customize them?
That of course requires that reading them in is dynamically instead of hardcoded.

yardenshoham commented:
Gotcha, so it'll go on `public`. Yeah, let's do that.

wxiaoguang commented:
I prefer LFS than bloating the codebase with hundreds (maybe thousands in the future) vendoring files.

lunny commented:
> I still don't like the introduction of LFS for this, it means the gitea repo can no longer be hosted on a default installation of gitea which has LFS disabled. 294kB is certainly not a size where I would even consider LFS, that starts in the realms of 100MB+ for me.
> 
> Not going to block if everyone else agrees to LFS, but I don't.

Maybe we could enable LFS default now. Because it will not require any extra installation or configuration for users.

JakobDev commented:
> Because it will not require any extra installation or configuration for users.

I think you need git LFS installed. LFS is not included in git by default.

lunny commented:
> > Because it will not require any extra installation or configuration for users.
> 
> I think you need git LFS installed. LFS is not included in git by default.



> > Because it will not require any extra installation or configuration for users.
> 
> I think you need git LFS installed. LFS is not included in git by default.

Don't need. Only git client need git-LFS. Gitea's LFS logic is written by Golang and it's internal.

delvh commented:
Isn't that the whole point of that argument?
We are not talking about Gitea but its users/developers.

lunny commented:
> Isn't that the whole point of that argument? We are not talking about Gitea but its users/developers.

For developers, they need to install git/nodejs/golang/python, I don't think installing git-lfs is difficult for them.

KN4CK3R commented:
I don't understand why we should want these icons in our codebase and inherit the burden of keeping them updated if we could rely on external icon packs.

denyskon commented:
I also don't think we need to store it in the codebase. Why not just download the icons when building?

lunny commented:
> I also don't think we need to store it in the codebase. Why not just download the icons when building?

Or download the icons when using. But it's more difficult. We need to define an icon package or repository specification.

silverwind commented:
I do think keeping build artifacts in the repo is beneficial so gitea can be built without internet access, we do that for numerous dependencies like Fomantic, the SVGs, etc.

silverwind commented:
We could add https://www.npmjs.com/package/material-icon-theme as a npm dependency, and build from `node_modules`, this is preferable to the current download script anyways because certain environments have custom npm registries and such where `registry.npmjs.org` is intentionally unreachable. By offloading the download to the npm client, we get this compatibilty for free, in addtion to having the dependency version-locked.

denyskon commented:
I think npm is the best option, even though we lose customizability. I don't think that many users will really want install a different icon theme.....

yardenshoham commented:
If there's no need for customizability we could have landed this months ago

silverwind commented:
We don't loose customizability just by changing the way the material icons are fetched.

We just need to formalize a spec format for a icon pack, e.g. that it should contain a JSON mapping file and the icons in a subdirectory. By the way, I find `zip` files more easy to handle than `tgz`, especially on Windows, so likely we should move to that format.

wxiaoguang commented:
I have no motivation for make more changes because [I do not like these colorful icons](https://github.com/go-gitea/gitea/pull/24147#issuecomment-1512811533). I wrote this PR just because [I promosed](https://github.com/go-gitea/gitea/pull/24147#issuecomment-1513343771).

While I do not think #24147 is good enough, eg: the packing was wrong (the [material-icons.json](https://github.com/go-gitea/gitea/pull/24147/files#diff-ebdc301b0804fd65b07d9a440e8268ca2558c91c5dc026e5aa06b2ec09a53843) shouldn't be assumed to be a local file or read by `os.ReadFile`)

And, mixing the external assets into Gitea's SVG assets doesn't look good to me.


Feel free to replace this PR, I would have no objection if the solution is maintainable and satisfies:

1. Not bloating the codebase too much
2. Not mixing the external assets into the Gitea's own assets

yardenshoham commented:
Thanks @wxiaoguang, it was worth a shot

wxiaoguang commented:
A new PR: 

-> Add material icons for file list #33837
